File: examples/tools/idalib_dump_port.cpp
/// \file idalib_dump_port.cpp
/// \brief idax-first port of `idalib-dump`'s `ida_dump` command.
///
/// This example intentionally avoids direct SDK calls and uses only idax public
/// wrappers. The goal is to validate real-world migration ergonomics and keep
/// parity pressure visible through concrete port code.

#include <ida/idax.hpp>

#include <algorithm>
#include <charconv>
#include <cctype>
#include <cstdio>
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <regex>
#include <sstream>
#include <string>
#include <string_view>
#include <unordered_set>
#include <utility>
#include <vector>

namespace {

template <typename... Args>
std::string fmt(const char* pattern, Args&&... args) {
    char buffer[4096];
    std::snprintf(buffer, sizeof(buffer), pattern, std::forward<Args>(args)...);
    return buffer;
}

std::string error_text(const ida::Error& error) {
    if (error.context.empty()) {
        return error.message;
    }
    return error.message + " (" + error.context + ")";
}

struct Options {
    std::string input_file;
    std::string output_file;
    std::string filter_pattern;
    std::vector<std::string> function_list;
    ida::Address filter_address{ida::BadAddress};

    bool show_assembly{true};
    bool show_pseudocode{true};
    bool show_microcode{false};
    bool list_only{false};
    bool errors_only{false};
    bool quiet{false};
    bool show_summary{true};

    // Compatibility switches from the original tool.
    bool no_plugins_requested{false};
    std::vector<std::string> plugin_patterns;
};

struct Stats {
    std::size_t total_functions{0};
    std::size_t selected_functions{0};
    std::size_t decompiled_ok{0};
    std::size_t decompiled_fail{0};
    std::size_t skipped{0};
    std::vector<std::pair<std::string, std::string>> errors;
};

struct FunctionNames {
    std::string raw;
    std::string demangled;
};

struct DecompileAttempt {
    bool attempted{false};
    bool success{false};
    std::vector<std::string> pseudocode_lines;
    std::vector<std::string> microcode_lines;
    std::string error;
    std::string microcode_error;
};

Options g_options;
Stats g_stats;
std::ostream* g_output = &std::cout;
std::ofstream g_output_file;

bool parse_hex_address_value(std::string_view text, ida::Address* out_value) {
    if (text.empty() || out_value == nullptr) {
        return false;
    }
    std::uint64_t parsed = 0;
    const char* begin = text.data();
    const char* end = begin + text.size();
    auto [ptr, ec] = std::from_chars(begin, end, parsed, 16);
    if (ec != std::errc{} || ptr != end) {
        return false;
    }
    *out_value = static_cast<ida::Address>(parsed);
    return true;
}

std::vector<std::string> split_list(std::string_view text) {
    std::vector<std::string> out;
    std::string current;
    for (char c : text) {
        if (c == ',' || c == '|') {
            if (!current.empty()) {
                auto first = current.find_first_not_of(" \t\r\n");
                auto last = current.find_last_not_of(" \t\r\n");
                if (first != std::string::npos) {
                    out.push_back(current.substr(first, last - first + 1));
                }
                current.clear();
            }
            continue;
        }
        current.push_back(c);
    }
    if (!current.empty()) {
        auto first = current.find_first_not_of(" \t\r\n");
        auto last = current.find_last_not_of(" \t\r\n");
        if (first != std::string::npos) {
            out.push_back(current.substr(first, last - first + 1));
        }
    }
    return out;
}

bool looks_like_generated_name(std::string_view text) {
    return text.rfind("sub_", 0) == 0
        || text.rfind("loc_", 0) == 0
        || text.rfind("unk_", 0) == 0
        || text.rfind("off_", 0) == 0
        || text.rfind("byte_", 0) == 0
        || text.rfind("word_", 0) == 0
        || text.rfind("dword_", 0) == 0
        || text.rfind("qword_", 0) == 0;
}

ida::Address parse_address(std::string_view text) {
    if (text.empty()) {
        return ida::BadAddress;
    }

    if (looks_like_generated_name(text)) {
        return ida::BadAddress;
    }

    std::string s(text);
    if (s.rfind("0x", 0) == 0 || s.rfind("0X", 0) == 0) {
        s = s.substr(2);
    }
    if (s.empty()) {
        return ida::BadAddress;
    }
    for (char c : s) {
        if (!std::isxdigit(static_cast<unsigned char>(c))) {
            return ida::BadAddress;
        }
    }
    ida::Address parsed = ida::BadAddress;
    return parse_hex_address_value(s, &parsed) ? parsed : ida::BadAddress;
}

bool regex_or_substring_match(const std::string& pattern, const std::string& candidate) {
    if (candidate.empty()) {
        return false;
    }
    try {
        std::regex re(pattern, std::regex::icase);
        return std::regex_search(candidate, re);
    } catch (const std::regex_error&) {
        std::string lowered_candidate = candidate;
        std::string lowered_pattern = pattern;
        std::transform(lowered_candidate.begin(), lowered_candidate.end(),
                       lowered_candidate.begin(),
                       [](unsigned char c) { return static_cast<char>(std::tolower(c)); });
        std::transform(lowered_pattern.begin(), lowered_pattern.end(),
                       lowered_pattern.begin(),
                       [](unsigned char c) { return static_cast<char>(std::tolower(c)); });
        return lowered_candidate.find(lowered_pattern) != std::string::npos;
    }
}

bool equals_case_insensitive(std::string_view a, std::string_view b) {
    if (a.size() != b.size()) {
        return false;
    }
    for (std::size_t i = 0; i < a.size(); ++i) {
        if (std::tolower(static_cast<unsigned char>(a[i]))
            != std::tolower(static_cast<unsigned char>(b[i]))) {
            return false;
        }
    }
    return true;
}

FunctionNames resolve_function_names(const ida::function::Function& function) {
    FunctionNames out;
    out.raw = function.name();
    if (auto demangled = ida::name::demangled(function.start(), ida::name::DemangleForm::Full)) {
        out.demangled = *demangled;
    }
    return out;
}

bool matches_explicit_list(const FunctionNames& names, ida::Address start_address) {
    if (g_options.function_list.empty()) {
        return true;
    }

    for (const auto& item : g_options.function_list) {
        ida::Address parsed = parse_address(item);
        if (parsed != ida::BadAddress) {
            if (parsed == start_address) {
                return true;
            }
            continue;
        }

        if (item == names.raw || item == names.demangled) {
            return true;
        }
        if (equals_case_insensitive(item, names.raw)
            || (!names.demangled.empty() && equals_case_insensitive(item, names.demangled))) {
            return true;
        }
    }

    return false;
}

bool should_process_function(const ida::function::Function& function,
                             const FunctionNames& names) {
    if (g_options.filter_address != ida::BadAddress && function.start() != g_options.filter_address) {
        return false;
    }

    if (!matches_explicit_list(names, function.start())) {
        return false;
    }

    if (!g_options.filter_pattern.empty()) {
        if (regex_or_substring_match(g_options.filter_pattern, names.raw)) {
            return true;
        }
        if (!names.demangled.empty()
            && regex_or_substring_match(g_options.filter_pattern, names.demangled)) {
            return true;
        }
        return false;
    }

    return true;
}

class DatabaseSession {
public:
    ida::Status open(std::string_view input_path) {
        ida::database::RuntimeOptions runtime_options;
        runtime_options.quiet = g_options.quiet;
        runtime_options.plugin_policy.disable_user_plugins = g_options.no_plugins_requested;
        runtime_options.plugin_policy.allowlist_patterns = g_options.plugin_patterns;

        if (auto init_status = ida::database::init(runtime_options); !init_status) {
            return std::unexpected(init_status.error());
        }
        if (auto open_status = ida::database::open(input_path, ida::database::OpenMode::Analyze);
            !open_status) {
            return std::unexpected(open_status.error());
        }
        is_open_ = true;

        if (auto wait_status = ida::analysis::wait(); !wait_status) {
            return std::unexpected(wait_status.error());
        }

        return ida::ok();
    }

    ~DatabaseSession() {
        if (is_open_) {
            ida::database::close(false);
        }
    }

private:
    bool is_open_{false};
};

void print_usage(const char* program) {
    std::cout << "idax_dump_port - idalib-dump port using idax wrappers\n\n";
    std::cout << "Usage: " << program << " [options] <binary_file>\n\n";
    std::cout << "Output selection:\n";
    std::cout << "  --asm             include assembly output\n";
    std::cout << "  --pseudo          include pseudocode output\n";
    std::cout << "  --mc              include microcode output\n";
    std::cout << "  --asm-only        show only assembly\n";
    std::cout << "  --pseudo-only     show only pseudocode\n";
    std::cout << "  --mc-only         show only microcode\n\n";
    std::cout << "Filtering:\n";
    std::cout << "  -f, --filter <pattern>   regex/substring function-name filter\n";
    std::cout << "  -F, --functions <list>   comma/pipe-separated names/addresses\n";
    std::cout << "  -a, --address <hex>      function entry address filter\n";
    std::cout << "  -e, --errors             only show decompilation failures\n";
    std::cout << "  -l, --list               list matching functions only\n\n";
    std::cout << "Output control:\n";
    std::cout << "  -o, --output <file>      write output to file\n";
    std::cout << "  -q, --quiet              suppress startup summary\n";
    std::cout << "  --no-summary             disable final summary\n\n";
    std::cout << "Headless plugin policy:\n";
    std::cout << "  --no-plugins             disable user plugins from IDAUSR\n";
    std::cout << "  --plugin <pattern>       allow only matching user plugins (* and ? supported)\n\n";
    std::cout << "  -h, --help               show this help\n";
}

bool parse_arguments(int argc, char* argv[]) {
    bool asm_selected = false;
    bool pseudo_selected = false;
    bool microcode_selected = false;

    for (int i = 1; i < argc; ++i) {
        std::string arg = argv[i];

        if (arg == "-h" || arg == "--help") {
            print_usage(argv[0]);
            std::exit(EXIT_SUCCESS);
        }
        if (arg == "-o" || arg == "--output") {
            if (i + 1 >= argc) {
                std::cerr << "--output requires a file path\n";
                return false;
            }
            g_options.output_file = argv[++i];
            continue;
        }
        if (arg == "-f" || arg == "--filter") {
            if (i + 1 >= argc) {
                std::cerr << "--filter requires a pattern\n";
                return false;
            }
            g_options.filter_pattern = argv[++i];
            continue;
        }
        if (arg == "-F" || arg == "--functions") {
            if (i + 1 >= argc) {
                std::cerr << "--functions requires a list\n";
                return false;
            }
            g_options.function_list = split_list(argv[++i]);
            continue;
        }
        if (arg == "-a" || arg == "--address") {
            if (i + 1 >= argc) {
                std::cerr << "--address requires a value\n";
                return false;
            }
            g_options.filter_address = parse_address(argv[++i]);
            if (g_options.filter_address == ida::BadAddress) {
                std::cerr << "invalid --address value\n";
                return false;
            }
            continue;
        }
        if (arg == "-e" || arg == "--errors") {
            g_options.errors_only = true;
            continue;
        }
        if (arg == "-l" || arg == "--list") {
            g_options.list_only = true;
            continue;
        }
        if (arg == "-q" || arg == "--quiet") {
            g_options.quiet = true;
            continue;
        }
        if (arg == "--no-summary") {
            g_options.show_summary = false;
            continue;
        }

        if (arg == "--asm") {
            asm_selected = true;
            continue;
        }
        if (arg == "--pseudo") {
            pseudo_selected = true;
            continue;
        }
        if (arg == "--mc") {
            microcode_selected = true;
            continue;
        }
        if (arg == "--asm-only") {
            g_options.show_assembly = true;
            g_options.show_pseudocode = false;
            g_options.show_microcode = false;
            continue;
        }
        if (arg == "--pseudo-only") {
            g_options.show_assembly = false;
            g_options.show_pseudocode = true;
            g_options.show_microcode = false;
            continue;
        }
        if (arg == "--mc-only") {
            g_options.show_assembly = false;
            g_options.show_pseudocode = false;
            g_options.show_microcode = true;
            microcode_selected = true;
            continue;
        }

        if (arg == "--no-plugins") {
            g_options.no_plugins_requested = true;
            continue;
        }
        if (arg == "--plugin") {
            if (i + 1 >= argc) {
                std::cerr << "--plugin requires a pattern\n";
                return false;
            }
            auto patterns = split_list(argv[++i]);
            g_options.plugin_patterns.insert(g_options.plugin_patterns.end(),
                                             patterns.begin(),
                                             patterns.end());
            g_options.no_plugins_requested = true;
            continue;
        }

        if (!arg.empty() && arg[0] == '-') {
            std::cerr << "unknown option: " << arg << "\n";
            return false;
        }

        if (g_options.input_file.empty()) {
            g_options.input_file = arg;
        } else {
            std::cerr << "multiple input files are not supported\n";
            return false;
        }
    }

    if (g_options.input_file.empty()) {
        std::cerr << "no input file provided\n";
        return false;
    }

    if (asm_selected || pseudo_selected || microcode_selected) {
        g_options.show_assembly = asm_selected;
        g_options.show_pseudocode = pseudo_selected;
        g_options.show_microcode = microcode_selected;
    }

    return true;
}

void print_header(const ida::function::Function& function,
                  const FunctionNames& names,
                  bool success,
                  std::string_view status_text = {}) {
    *g_output << "\n" << std::string(78, '=') << "\n";
    *g_output << fmt("Function: %s (%#llx) [%s]",
                     names.raw.c_str(),
                     static_cast<unsigned long long>(function.start()),
                     success ? "OK" : "FAIL");
    if (!status_text.empty()) {
        *g_output << " - " << status_text;
    }
    *g_output << "\n" << std::string(78, '-') << "\n";
}

void dump_assembly(const ida::function::Function& function) {
    *g_output << "-- Assembly " << std::string(66, '-') << "\n";

    auto addresses = ida::function::code_addresses(function.start());
    if (!addresses) {
        *g_output << "  [error] could not enumerate instruction addresses: "
                  << error_text(addresses.error()) << "\n\n";
        return;
    }

    for (ida::Address address : *addresses) {
        auto text = ida::instruction::text(address);
        if (!text) {
            *g_output << fmt("%#llx  <decode failed>\n",
                             static_cast<unsigned long long>(address));
            continue;
        }
        *g_output << fmt("%#llx  %s\n",
                         static_cast<unsigned long long>(address),
                         text->c_str());
    }
    *g_output << "\n";
}

DecompileAttempt try_decompile(const ida::function::Function& function,
                               bool decompiler_available,
                               bool need_pseudocode,
                               bool need_microcode) {
    DecompileAttempt attempt;
    if (!decompiler_available) {
        attempt.error = "Hex-Rays decompiler unavailable";
        return attempt;
    }

    attempt.attempted = true;

    ida::decompiler::DecompileFailure failure;
    auto decompiled = ida::decompiler::decompile(function.start(), &failure);
    if (!decompiled) {
        attempt.error = error_text(decompiled.error());
        if (!failure.description.empty()) {
            attempt.error += fmt(" [failure_ea=%#llx: %s]",
                                 static_cast<unsigned long long>(failure.failure_address),
                                 failure.description.c_str());
        }
        return attempt;
    }

    if (need_pseudocode) {
        auto lines = decompiled->lines();
        if (!lines) {
            attempt.error = error_text(lines.error());
            return attempt;
        }
        attempt.pseudocode_lines = std::move(*lines);
    }

    if (need_microcode) {
        auto mc_lines = decompiled->microcode_lines();
        if (mc_lines) {
            attempt.microcode_lines = std::move(*mc_lines);
        } else {
            attempt.microcode_error = error_text(mc_lines.error());
        }
    }

    attempt.success = true;
    return attempt;
}

void dump_pseudocode_lines(const std::vector<std::string>& lines) {
    *g_output << "-- Pseudocode " << std::string(65, '-') << "\n";
    for (const auto& line : lines) {
        *g_output << line << "\n";
    }
    *g_output << "\n";
}

void dump_microcode_lines(const std::vector<std::string>& lines) {
    *g_output << "-- Microcode " << std::string(66, '-') << "\n";
    for (const auto& line : lines) {
        *g_output << line << "\n";
    }
    *g_output << "\n";
}

void print_startup_metadata() {
    if (g_options.quiet) {
        return;
    }
    *g_output << "idax dump port\n";
    if (auto path = ida::database::input_file_path()) {
        *g_output << "  Input: " << *path << "\n";
    }
    if (auto file_type = ida::database::file_type_name()) {
        *g_output << "  File type: " << *file_type << "\n";
    }
    if (auto loader_format = ida::database::loader_format_name()) {
        *g_output << "  Loader format: " << *loader_format << "\n";
    }
    if (auto md5 = ida::database::input_md5()) {
        *g_output << "  MD5:   " << *md5 << "\n";
    }
    if (auto compiler = ida::database::compiler_info()) {
        *g_output << "  Compiler: " << compiler->name;
        if (!compiler->abbreviation.empty()) {
            *g_output << " (" << compiler->abbreviation << ")";
        }
        if (compiler->uncertain) {
            *g_output << " [uncertain]";
        }
        *g_output << "\n";
    }
    if (auto modules = ida::database::import_modules()) {
        std::size_t symbol_count = 0;
        for (const auto& module : *modules) {
            symbol_count += module.symbols.size();
        }
        *g_output << "  Import modules: " << modules->size()
                  << " (symbols: " << symbol_count << ")\n";
    }
    if (auto min = ida::database::min_address(); min) {
        if (auto max = ida::database::max_address(); max) {
            *g_output << fmt("  Range: %#llx - %#llx\n",
                             static_cast<unsigned long long>(*min),
                             static_cast<unsigned long long>(*max));
        }
    }
    *g_output << "\n";
}

void print_summary() {
    if (!g_options.show_summary || g_options.list_only) {
        return;
    }

    std::cout << "\n" << std::string(78, '=') << "\n";
    std::cout << "Summary\n";
    std::cout << std::string(78, '-') << "\n";
    std::cout << "  Total functions:    " << g_stats.total_functions << "\n";
    std::cout << "  Selected:           " << g_stats.selected_functions << "\n";
    std::cout << "  Decompiled OK:      " << g_stats.decompiled_ok << "\n";
    std::cout << "  Decompile failures: " << g_stats.decompiled_fail << "\n";
    std::cout << "  Skipped:            " << g_stats.skipped << "\n";

    if (!g_stats.errors.empty()) {
        std::cout << "\nErrors:\n";
        for (const auto& [name, error] : g_stats.errors) {
            std::cout << "  " << name << ": " << error << "\n";
        }
    }

    std::cout << std::string(78, '=') << "\n";
}

int run_port() {
    if (!g_options.output_file.empty()) {
        g_output_file.open(g_options.output_file);
        if (!g_output_file.is_open()) {
            std::cerr << "failed to open output file: " << g_options.output_file << "\n";
            return EXIT_FAILURE;
        }
        g_output = &g_output_file;
    }

    DatabaseSession session;
    if (auto open_status = session.open(g_options.input_file); !open_status) {
        std::cerr << "failed to initialize analysis session: "
                  << error_text(open_status.error()) << "\n";
        return EXIT_FAILURE;
    }

    print_startup_metadata();

    bool decompiler_available = false;
    if (auto available = ida::decompiler::available(); available && *available) {
        decompiler_available = true;
    }

    if ((g_options.show_pseudocode || g_options.show_microcode || g_options.errors_only)
        && !decompiler_available) {
        std::cerr << "[warning] decompiler unavailable; pseudocode/microcode/error filtering disabled\n";
        if (g_options.errors_only) {
            std::cerr << "[error] --errors requires decompiler availability\n";
            return EXIT_FAILURE;
        }
    }

    auto function_count = ida::function::count();
    if (!function_count) {
        std::cerr << "failed to enumerate functions: "
                  << error_text(function_count.error()) << "\n";
        return EXIT_FAILURE;
    }
    g_stats.total_functions = *function_count;

    if (g_options.list_only) {
        *g_output << fmt("%-18s %-8s %s\n", "Address", "Size", "Name");
        *g_output << std::string(78, '-') << "\n";
    }

    for (auto function : ida::function::all()) {
        FunctionNames names = resolve_function_names(function);
        if (!should_process_function(function, names)) {
            ++g_stats.skipped;
            continue;
        }

        ++g_stats.selected_functions;

        if (g_options.list_only) {
            *g_output << fmt("%#-16llx %-8llu %s\n",
                             static_cast<unsigned long long>(function.start()),
                             static_cast<unsigned long long>(function.size()),
                             names.raw.c_str());
            continue;
        }

        DecompileAttempt decompile;
        if (g_options.show_pseudocode || g_options.show_microcode || g_options.errors_only) {
            decompile = try_decompile(function,
                                      decompiler_available,
                                      g_options.show_pseudocode,
                                      g_options.show_microcode);
        }

        if (g_options.errors_only) {
            if (!decompile.attempted || decompile.success) {
                ++g_stats.skipped;
                continue;
            }
        }

        if (decompile.attempted && decompile.success) {
            ++g_stats.decompiled_ok;
            print_header(function, names, true);
        } else if (decompile.attempted && !decompile.success) {
            ++g_stats.decompiled_fail;
            g_stats.errors.emplace_back(names.raw, decompile.error);
            print_header(function, names, false, decompile.error);
        } else {
            print_header(function, names, true);
        }

        if (g_options.show_assembly) {
            dump_assembly(function);
        }

        if (g_options.show_pseudocode) {
            if (decompile.attempted && decompile.success) {
                dump_pseudocode_lines(decompile.pseudocode_lines);
            } else {
                *g_output << "-- Pseudocode " << std::string(65, '-') << "\n";
                *g_output << "  [unavailable] "
                          << (decompile.error.empty() ? "decompilation unavailable" : decompile.error)
                          << "\n\n";
            }
        }

        if (g_options.show_microcode) {
            if (decompile.attempted && decompile.success
                && !decompile.microcode_lines.empty()) {
                dump_microcode_lines(decompile.microcode_lines);
            } else {
                *g_output << "-- Microcode " << std::string(66, '-') << "\n";
                const std::string reason = decompile.microcode_error.empty()
                    ? (decompile.error.empty() ? "microcode unavailable" : decompile.error)
                    : decompile.microcode_error;
                *g_output << "  [unavailable] " << reason << "\n\n";
            }
        }
    }

    print_summary();
    return g_stats.decompiled_fail == 0 ? EXIT_SUCCESS : EXIT_FAILURE;
}

} // namespace

int main(int argc, char* argv[]) {
    if (!parse_arguments(argc, argv)) {
        std::cerr << "Use --help for usage.\n";
        return EXIT_FAILURE;
    }
    return run_port();
}

File: bindings/rust/idax/examples/idalib_dump_port.rs
mod common;

use common::{DatabaseSession, format_error, print_usage, write_output};
use idax::{Error, Result, database, decompiler, function, instruction};

#[derive(Debug, Clone)]
struct Options {
    input: String,
    output: Option<String>,
    filter: Option<String>,
    function_names: Vec<String>,
    list_only: bool,
    show_assembly: bool,
    show_pseudocode: bool,
    show_microcode: bool,
    no_summary: bool,
    max_asm_lines: usize,
}

impl Default for Options {
    fn default() -> Self {
        Self {
            input: String::new(),
            output: None,
            filter: None,
            function_names: Vec::new(),
            list_only: false,
            show_assembly: true,
            show_pseudocode: true,
            show_microcode: false,
            no_summary: false,
            max_asm_lines: 120,
        }
    }
}

fn parse_options(args: &[String]) -> Result<Options> {
    if args.len() < 2 {
        return Err(Error::validation("missing binary_file argument"));
    }

    let mut options = Options {
        input: args[1].clone(),
        ..Options::default()
    };

    let mut index = 2usize;
    while index < args.len() {
        match args[index].as_str() {
            "-h" | "--help" => {
                print_usage(
                    &args[0],
                    "<binary_file> [--list] [--asm] [--pseudo] [--microcode] [--asm-only] [--pseudo-only] [--microcode-only] [--filter <text>] [--function <name>] [--output <path>] [--max-asm-lines <count>] [--no-summary]",
                );
                std::process::exit(0);
            }
            "--list" | "-l" => {
                options.list_only = true;
            }
            "--asm" => options.show_assembly = true,
            "--pseudo" => options.show_pseudocode = true,
            "--microcode" | "--mc" => options.show_microcode = true,
            "--asm-only" => {
                options.show_assembly = true;
                options.show_pseudocode = false;
                options.show_microcode = false;
            }
            "--pseudo-only" => {
                options.show_assembly = false;
                options.show_pseudocode = true;
                options.show_microcode = false;
            }
            "--microcode-only" | "--mc-only" => {
                options.show_assembly = false;
                options.show_pseudocode = false;
                options.show_microcode = true;
            }
            "--filter" | "-f" => {
                index += 1;
                if index >= args.len() {
                    return Err(Error::validation("--filter requires a value"));
                }
                options.filter = Some(args[index].clone());
            }
            "--function" | "-F" => {
                index += 1;
                if index >= args.len() {
                    return Err(Error::validation("--function requires a value"));
                }
                options.function_names.push(args[index].clone());
            }
            "--output" | "-o" => {
                index += 1;
                if index >= args.len() {
                    return Err(Error::validation("--output requires a path"));
                }
                options.output = Some(args[index].clone());
            }
            "--max-asm-lines" => {
                index += 1;
                if index >= args.len() {
                    return Err(Error::validation("--max-asm-lines requires a value"));
                }
                options.max_asm_lines = args[index]
                    .parse::<usize>()
                    .map_err(|_| Error::validation("invalid --max-asm-lines value"))?;
            }
            "--no-summary" => options.no_summary = true,
            unknown => {
                return Err(Error::validation(format!("unknown option: {unknown}")));
            }
        }
        index += 1;
    }

    if !options.list_only
        && !options.show_assembly
        && !options.show_pseudocode
        && !options.show_microcode
    {
        options.show_assembly = true;
        options.show_pseudocode = true;
    }

    Ok(options)
}

fn function_matches(function_name: &str, options: &Options) -> bool {
    if !options.function_names.is_empty() {
        let exact = options
            .function_names
            .iter()
            .any(|candidate| candidate == function_name);
        if !exact {
            return false;
        }
    }
    if let Some(pattern) = &options.filter {
        return function_name.contains(pattern);
    }
    true
}

fn format_hex(address: u64) -> String {
    format!("0x{address:x}")
}

fn run() -> Result<()> {
    let args: Vec<String> = std::env::args().collect();
    let options = parse_options(&args)?;

    let _session = DatabaseSession::open(&options.input, true)?;
    let decompiler_available = decompiler::available().unwrap_or(false);

    let mut selected = Vec::new();
    for function in function::all() {
        if function_matches(function.name(), &options) {
            selected.push(function);
        }
    }

    let mut output = String::new();
    let mut decompile_failures: Vec<(u64, String, String)> = Vec::new();

    if options.list_only {
        output.push_str("Address              Size      Name\n");
        output.push_str("---------------------------------------------\n");
        for function in &selected {
            output.push_str(&format!(
                "{:<20} {:<9} {}\n",
                format_hex(function.start()),
                function.size(),
                function.name()
            ));
        }
    } else {
        for function in &selected {
            output.push_str(&format!(
                "============================================================\nFunction: {} @ {} (size={})\n============================================================\n",
                function.name(),
                format_hex(function.start()),
                function.size()
            ));

            if options.show_assembly {
                output.push_str("\n-- Assembly --\n");
                let addresses = function::code_addresses(function.start())?;
                for (index, address) in addresses.iter().take(options.max_asm_lines).enumerate() {
                    match instruction::text(*address) {
                        Ok(text) => {
                            output.push_str(&format!(
                                "{:04}  {}  {}\n",
                                index,
                                format_hex(*address),
                                text
                            ));
                        }
                        Err(error) => {
                            output.push_str(&format!(
                                "{:04}  {}  <decode error: {}>\n",
                                index,
                                format_hex(*address),
                                format_error(&error)
                            ));
                        }
                    }
                }
            }

            if options.show_pseudocode || options.show_microcode {
                if decompiler_available {
                    match decompiler::decompile(function.start()) {
                        Ok(decompiled) => {
                            if options.show_pseudocode {
                                output.push_str("\n-- Pseudocode --\n");
                                match decompiled.pseudocode() {
                                    Ok(text) => {
                                        output.push_str(&text);
                                        output.push('\n');
                                    }
                                    Err(error) => {
                                        output.push_str(&format!(
                                            "<pseudocode error: {}>\n",
                                            format_error(&error)
                                        ));
                                    }
                                }
                            }
                            if options.show_microcode {
                                output.push_str("\n-- Microcode --\n");
                                match decompiled.microcode() {
                                    Ok(text) => {
                                        output.push_str(&text);
                                        output.push('\n');
                                    }
                                    Err(error) => {
                                        output.push_str(&format!(
                                            "<microcode error: {}>\n",
                                            format_error(&error)
                                        ));
                                    }
                                }
                            }
                        }
                        Err(error) => {
                            decompile_failures.push((
                                function.start(),
                                function.name().to_string(),
                                format_error(&error),
                            ));
                        }
                    }
                } else {
                    output.push_str("\n-- Decompiler --\n<Hex-Rays unavailable on this host>\n");
                }
            }

            output.push('\n');
        }
    }

    if !options.no_summary {
        let total_functions = function::count()?;
        output.push_str("\n================ Summary ================\n");
        output.push_str(&format!(
            "Input: {}\nTotal functions: {}\nSelected functions: {}\nDecompiler failures: {}\n",
            options.input,
            total_functions,
            selected.len(),
            decompile_failures.len(),
        ));
        if !decompile_failures.is_empty() {
            output.push_str("\nDecompiler failures:\n");
            for (address, name, reason) in &decompile_failures {
                output.push_str(&format!(
                    "  - {} {}: {}\n",
                    format_hex(*address),
                    name,
                    reason
                ));
            }
        }
    }

    write_output(options.output.as_deref(), &output)?;

    if !decompile_failures.is_empty() {
        return Err(Error::sdk(format!(
            "{} function(s) failed to decompile",
            decompile_failures.len()
        )));
    }

    Ok(())
}

fn main() {
    if let Err(error) = run() {
        eprintln!("error: {}", format_error(&error));
        std::process::exit(1);
    }
}
